In [2]:
import pandas as pd
import numpy as np


Website Dataframe info

https://archive.ics.uci.edu/dataset/891/cdc+diabetes+health+indicators

In [7]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
cdc_diabetes_health_indicators = fetch_ucirepo(id=891) 
  
# data (as pandas dataframes) 
X = cdc_diabetes_health_indicators.data.features 
y = cdc_diabetes_health_indicators.data.targets 
  
# metadata 
print(cdc_diabetes_health_indicators.metadata) 
  
# variable information 
print(cdc_diabetes_health_indicators.variables) 
pd.set_option('display.max_row', None)
pd.set_option('display.max_column', None)

{'uci_id': 891, 'name': 'CDC Diabetes Health Indicators', 'repository_url': 'https://archive.ics.uci.edu/dataset/891/cdc+diabetes+health+indicators', 'data_url': 'https://archive.ics.uci.edu/static/public/891/data.csv', 'abstract': 'The Diabetes Health Indicators Dataset contains healthcare statistics and lifestyle survey information about people in general along with their diagnosis of diabetes. The 35 features consist of some demographics, lab test results, and answers to survey questions for each patient. The target variable for classification is whether a patient has diabetes, is pre-diabetic, or healthy. ', 'area': 'Life Science', 'tasks': ['Classification'], 'characteristics': ['Tabular', 'Multivariate'], 'num_instances': 253680, 'num_features': 21, 'feature_types': ['Categorical', 'Integer'], 'demographics': ['Sex', 'Age', 'Education Level', 'Income'], 'target_col': ['Diabetes_binary'], 'index_col': ['ID'], 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_data

In [8]:
X.head()

HighBP  HighChol  CholCheck  BMI  Smoker  Stroke  HeartDiseaseorAttack   
0       1         1          1   40       1       0                     0  \
1       0         0          0   25       1       0                     0   
2       1         1          1   28       0       0                     0   
3       1         0          1   27       0       0                     0   
4       1         1          1   24       0       0                     0   

   PhysActivity  Fruits  Veggies  HvyAlcoholConsump  AnyHealthcare   
0             0       0        1                  0              1  \
1             1       0        0                  0              0   
2             0       1        0                  0              1   
3             1       1        1                  0              1   
4             1       1        1                  0              1   

   NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  Age  Education   
0            0        5        18        15         1    0    9          4  \
1            1        3         0         0         0    0    7          6   
2            1        5        30        30         1    0    9          4   
3            0        2         0         0         0    0   11          3   
4            0        2         3         0         0    0   11          5   

   Income  
0       3  
1       1  
2       8  
3       6  
4       4

In [9]:
y.value_counts()

Diabetes_binary
0                  218334
1                   35346
Name: count, dtype: int64

In [8]:
X.shape

(253680, 21)

In [10]:
def check_columns(df):
    """
    This function takes a pandas dataframe as input and returns
    a dataframe with information about each column in the dataframe. For
    each column, it returns the column name, the number of
    unique values in the column, the unique values themselves,
    the number of null values in the column, and the data type of the column.
    The resulting dataframe is sorted by the 'Number of Unique Values' column in ascending order.
​
    Args:
    - df: pandas dataframe
​
    Returns:
    - pandas dataframe
    """
    data = []
    # Loop through each column in the dataframe
    for column in df.columns:
        # Append the column name, number of unique values, unique values, number of null values, and data type to the data list
        data.append(
            [
                column,
                df[column].nunique(),
                df[column].unique(),
                df[column].isna().sum(),
                df[column].isna().mean(),
                df[column].dtype
            ]
        )
    # Create a pandas dataframe from the data list, with column names 'Column Name', 'Number of Unique Values', 'Unique Values', 'Number of Null Values', and 'dtype'
    # Sort the resulting dataframe by the 'Number of Unique Values' column in ascending order
    return pd.DataFrame(
        data,
        columns=[
            "Column Name",
            "Number of Unique Values",
            "Unique Values",
            "Number of Null Values",
            "Proportion of Null Values",
            "dtype"
        ],
    ).sort_values(by="Number of Unique Values")

In [11]:
check_columns(X)

Column Name  Number of Unique Values   
0                 HighBP                        2  \
17                   Sex                        2   
16              DiffWalk                        2   
12           NoDocbcCost                        2   
11         AnyHealthcare                        2   
9                Veggies                        2   
8                 Fruits                        2   
10     HvyAlcoholConsump                        2   
6   HeartDiseaseorAttack                        2   
5                 Stroke                        2   
4                 Smoker                        2   
2              CholCheck                        2   
1               HighChol                        2   
7           PhysActivity                        2   
13               GenHlth                        5   
19             Education                        6   
20                Income                        8   
18                   Age                       13   
14              MentHlth                       31   
15              PhysHlth                       31   
3                    BMI                       84   

                                        Unique Values  Number of Null Values   
0                                              [1, 0]                      0  \
17                                             [0, 1]                      0   
16                                             [1, 0]                      0   
12                                             [0, 1]                      0   
11                                             [1, 0]                      0   
9                                              [1, 0]                      0   
8                                              [0, 1]                      0   
10                                             [0, 1]                      0   
6                                              [0, 1]                      0   
5                                              [0, 1]                      0   
4                                              [1, 0]                      0   
2                                              [1, 0]                      0   
1                                              [1, 0]                      0   
7                                              [0, 1]                      0   
13                                    [5, 3, 2, 4, 1]                      0   
19                                 [4, 6, 3, 5, 2, 1]                      0   
20                           [3, 1, 8, 6, 4, 7, 2, 5]                      0   
18        [9, 7, 11, 10, 8, 13, 4, 6, 2, 12, 5, 1, 3]                      0   
14  [18, 0, 30, 3, 5, 15, 10, 6, 20, 2, 25, 1, 4, ...                      0   
15  [15, 0, 30, 2, 14, 28, 7, 20, 3, 10, 1, 5, 17,...                      0   
3   [40, 25, 28, 27, 24, 30, 34, 26, 33, 21, 23, 2...                      0   

    Proportion of Null Values  dtype  
0                         0.0  int64  
17                        0.0  int64  
16                        0.0  int64  
12                        0.0  int64  
11                        0.0  int64  
9                         0.0  int64  
8                         0.0  int64  
10                        0.0  int64  
6                         0.0  int64  
5                         0.0  int64  
4                         0.0  int64  
2                         0.0  int64  
1                         0.0  int64  
7                         0.0  int64  
13                        0.0  int64  
19                        0.0  int64  
20                        0.0  int64  
18                        0.0  int64  
14                        0.0  int64  
15                        0.0  int64  
3                         0.0  int64

https://www.kaggle.com/datasets/alexteboul/diabetes-health-indicators-dataset/data?select=diabetes_012_health_indicators_BRFSS2015.csv


### 1. diabetes _ 012 _ health _ indicators _ BRFSS2015.csv is a clean dataset of 253,680 survey responses to the CDC's BRFSS2015. The target variable Diabetes_012 has 3 classes. 0 is for no diabetes or only during pregnancy, 1 is for prediabetes, and 2 is for diabetes. There is class imbalance in this dataset. This dataset has 21 feature variables


In [12]:
diabetes_df1 = pd.read_csv('diabetes_012_health_indicators_BRFSS2015.csv')

In [14]:
X.head()

HighBP  HighChol  CholCheck  BMI  Smoker  Stroke  HeartDiseaseorAttack   
0       1         1          1   40       1       0                     0  \
1       0         0          0   25       1       0                     0   
2       1         1          1   28       0       0                     0   
3       1         0          1   27       0       0                     0   
4       1         1          1   24       0       0                     0   

   PhysActivity  Fruits  Veggies  HvyAlcoholConsump  AnyHealthcare   
0             0       0        1                  0              1  \
1             1       0        0                  0              0   
2             0       1        0                  0              1   
3             1       1        1                  0              1   
4             1       1        1                  0              1   

   NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  Age  Education   
0            0        5        18        15         1    0    9          4  \
1            1        3         0         0         0    0    7          6   
2            1        5        30        30         1    0    9          4   
3            0        2         0         0         0    0   11          3   
4            0        2         3         0         0    0   11          5   

   Income  
0       3  
1       1  
2       8  
3       6  
4       4

In [16]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253680 entries, 0 to 253679
Data columns (total 21 columns):
 #   Column                Non-Null Count   Dtype
---  ------                --------------   -----
 0   HighBP                253680 non-null  int64
 1   HighChol              253680 non-null  int64
 2   CholCheck             253680 non-null  int64
 3   BMI                   253680 non-null  int64
 4   Smoker                253680 non-null  int64
 5   Stroke                253680 non-null  int64
 6   HeartDiseaseorAttack  253680 non-null  int64
 7   PhysActivity          253680 non-null  int64
 8   Fruits                253680 non-null  int64
 9   Veggies               253680 non-null  int64
 10  HvyAlcoholConsump     253680 non-null  int64
 11  AnyHealthcare         253680 non-null  int64
 12  NoDocbcCost           253680 non-null  int64
 13  GenHlth               253680 non-null  int64
 14  MentHlth              253680 non-null  int64
 15  PhysHlth              253680 non-n

In [13]:
diabetes_df1.head()

Diabetes_012  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke   
0           0.0     1.0       1.0        1.0  40.0     1.0     0.0  \
1           0.0     0.0       0.0        0.0  25.0     1.0     0.0   
2           0.0     1.0       1.0        1.0  28.0     0.0     0.0   
3           0.0     1.0       0.0        1.0  27.0     0.0     0.0   
4           0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump   
0                   0.0           0.0     0.0      1.0                0.0  \
1                   0.0           1.0     0.0      0.0                0.0   
2                   0.0           0.0     1.0      0.0                0.0   
3                   0.0           1.0     1.0      1.0                0.0   
4                   0.0           1.0     1.0      1.0                0.0   

   AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   
0            1.0          0.0      5.0      18.0      15.0       1.0  0.0  \
1            0.0          1.0      3.0       0.0       0.0       0.0  0.0   
2            1.0          1.0      5.0      30.0      30.0       1.0  0.0   
3            1.0          0.0      2.0       0.0       0.0       0.0  0.0   
4            1.0          0.0      2.0       3.0       0.0       0.0  0.0   

    Age  Education  Income  
0   9.0        4.0     3.0  
1   7.0        6.0     1.0  
2   9.0        4.0     8.0  
3  11.0        3.0     6.0  
4  11.0        5.0     4.0

In [15]:
diabetes_df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253680 entries, 0 to 253679
Data columns (total 22 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Diabetes_012          253680 non-null  float64
 1   HighBP                253680 non-null  float64
 2   HighChol              253680 non-null  float64
 3   CholCheck             253680 non-null  float64
 4   BMI                   253680 non-null  float64
 5   Smoker                253680 non-null  float64
 6   Stroke                253680 non-null  float64
 7   HeartDiseaseorAttack  253680 non-null  float64
 8   PhysActivity          253680 non-null  float64
 9   Fruits                253680 non-null  float64
 10  Veggies               253680 non-null  float64
 11  HvyAlcoholConsump     253680 non-null  float64
 12  AnyHealthcare         253680 non-null  float64
 13  NoDocbcCost           253680 non-null  float64
 14  GenHlth               253680 non-null  float64
 15  

# X df and diabetes_df1 are the same, just diff dtype:   USE DIABETES_DF1 (it has diabetes cat already appended to it)

### 2. diabetes _ binary _ 5050split _ health _ indicators _ BRFSS2015.csv is a clean dataset of 70,692 survey responses to the CDC's BRFSS2015. It has an equal 50-50 split of respondents with no diabetes and with either prediabetes or diabetes. The target variable Diabetes_binary has 2 classes. 0 is for no diabetes, and 1 is for prediabetes or diabetes. This dataset has 21 feature variables and is balanced.


In [17]:
diabetes_50_50 = pd.read_csv('diabetes_binary_5050split_health_indicators_BRFSS2015.csv')

In [18]:
diabetes_50_50.head()

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke   
0              0.0     1.0       0.0        1.0  26.0     0.0     0.0  \
1              0.0     1.0       1.0        1.0  26.0     1.0     1.0   
2              0.0     0.0       0.0        1.0  26.0     0.0     0.0   
3              0.0     1.0       1.0        1.0  28.0     1.0     0.0   
4              0.0     0.0       0.0        1.0  29.0     1.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump   
0                   0.0           1.0     0.0      1.0                0.0  \
1                   0.0           0.0     1.0      0.0                0.0   
2                   0.0           1.0     1.0      1.0                0.0   
3                   0.0           1.0     1.0      1.0                0.0   
4                   0.0           1.0     1.0      1.0                0.0   

   AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   
0            1.0          0.0      3.0       5.0      30.0       0.0  1.0  \
1            1.0          0.0      3.0       0.0       0.0       0.0  1.0   
2            1.0          0.0      1.0       0.0      10.0       0.0  1.0   
3            1.0          0.0      3.0       0.0       3.0       0.0  1.0   
4            1.0          0.0      2.0       0.0       0.0       0.0  0.0   

    Age  Education  Income  
0   4.0        6.0     8.0  
1  12.0        6.0     8.0  
2  13.0        6.0     8.0  
3  11.0        6.0     8.0  
4   8.0        5.0     8.0

In [19]:
diabetes_50_50.shape

(70692, 22)

### 3. diabetes _ binary _ health _ indicators _ BRFSS2015.csv is a clean dataset of 253,680 survey responses to the CDC's BRFSS2015. The target variable Diabetes_binary has 2 classes. 0 is for no diabetes, and 1 is for prediabetes or diabetes. This dataset has 21 feature variables and is not balanced.

In [20]:
diabetes_df2 = pd.read_csv('diabetes_binary_health_indicators_BRFSS2015.csv')

In [21]:
diabetes_df2.head()

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke   
0              0.0     1.0       1.0        1.0  40.0     1.0     0.0  \
1              0.0     0.0       0.0        0.0  25.0     1.0     0.0   
2              0.0     1.0       1.0        1.0  28.0     0.0     0.0   
3              0.0     1.0       0.0        1.0  27.0     0.0     0.0   
4              0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  Veggies  HvyAlcoholConsump   
0                   0.0           0.0     0.0      1.0                0.0  \
1                   0.0           1.0     0.0      0.0                0.0   
2                   0.0           0.0     1.0      0.0                0.0   
3                   0.0           1.0     1.0      1.0                0.0   
4                   0.0           1.0     1.0      1.0                0.0   

   AnyHealthcare  NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   
0            1.0          0.0      5.0      18.0      15.0       1.0  0.0  \
1            0.0          1.0      3.0       0.0       0.0       0.0  0.0   
2            1.0          1.0      5.0      30.0      30.0       1.0  0.0   
3            1.0          0.0      2.0       0.0       0.0       0.0  0.0   
4            1.0          0.0      2.0       3.0       0.0       0.0  0.0   

    Age  Education  Income  
0   9.0        4.0     3.0  
1   7.0        6.0     1.0  
2   9.0        4.0     8.0  
3  11.0        3.0     6.0  
4  11.0        5.0     4.0

In [22]:
check_columns(diabetes_df2)

Column Name  Number of Unique Values   
0        Diabetes_binary                        2  \
18                   Sex                        2   
17              DiffWalk                        2   
13           NoDocbcCost                        2   
12         AnyHealthcare                        2   
11     HvyAlcoholConsump                        2   
9                 Fruits                        2   
8           PhysActivity                        2   
10               Veggies                        2   
6                 Stroke                        2   
5                 Smoker                        2   
3              CholCheck                        2   
2               HighChol                        2   
1                 HighBP                        2   
7   HeartDiseaseorAttack                        2   
14               GenHlth                        5   
20             Education                        6   
21                Income                        8   
19                   Age                       13   
15              MentHlth                       31   
16              PhysHlth                       31   
4                    BMI                       84   

                                        Unique Values  Number of Null Values   
0                                          [0.0, 1.0]                      0  \
18                                         [0.0, 1.0]                      0   
17                                         [1.0, 0.0]                      0   
13                                         [0.0, 1.0]                      0   
12                                         [1.0, 0.0]                      0   
11                                         [0.0, 1.0]                      0   
9                                          [0.0, 1.0]                      0   
8                                          [0.0, 1.0]                      0   
10                                         [1.0, 0.0]                      0   
6                                          [0.0, 1.0]                      0   
5                                          [1.0, 0.0]                      0   
3                                          [1.0, 0.0]                      0   
2                                          [1.0, 0.0]                      0   
1                                          [1.0, 0.0]                      0   
7                                          [0.0, 1.0]                      0   
14                          [5.0, 3.0, 2.0, 4.0, 1.0]                      0   
20                     [4.0, 6.0, 3.0, 5.0, 2.0, 1.0]                      0   
21           [3.0, 1.0, 8.0, 6.0, 4.0, 7.0, 2.0, 5.0]                      0   
19  [9.0, 7.0, 11.0, 10.0, 8.0, 13.0, 4.0, 6.0, 2....                      0   
15  [18.0, 0.0, 30.0, 3.0, 5.0, 15.0, 10.0, 6.0, 2...                      0   
16  [15.0, 0.0, 30.0, 2.0, 14.0, 28.0, 7.0, 20.0, ...                      0   
4   [40.0, 25.0, 28.0, 27.0, 24.0, 30.0, 34.0, 26....                      0   

    Proportion of Null Values    dtype  
0                         0.0  float64  
18                        0.0  float64  
17                        0.0  float64  
13                        0.0  float64  
12                        0.0  float64  
11                        0.0  float64  
9                         0.0  float64  
8                         0.0  float64  
10                        0.0  float64  
6                         0.0  float64  
5                         0.0  float64  
3                         0.0  float64  
2                         0.0  float64  
1                         0.0  float64  
7                         0.0  float64  
14                        0.0  float64  
20                        0.0  float64  
21                        0.0  float64  
19                        0.0  float64  
15                        0.0  float64  
16                        0.0  float64  
4                         0.0  float64

In [23]:
check_columns(diabetes_df1)

Column Name  Number of Unique Values   
10               Veggies                        2  \
1                 HighBP                        2   
2               HighChol                        2   
3              CholCheck                        2   
18                   Sex                        2   
5                 Smoker                        2   
6                 Stroke                        2   
7   HeartDiseaseorAttack                        2   
8           PhysActivity                        2   
9                 Fruits                        2   
17              DiffWalk                        2   
11     HvyAlcoholConsump                        2   
12         AnyHealthcare                        2   
13           NoDocbcCost                        2   
0           Diabetes_012                        3   
14               GenHlth                        5   
20             Education                        6   
21                Income                        8   
19                   Age                       13   
15              MentHlth                       31   
16              PhysHlth                       31   
4                    BMI                       84   

                                        Unique Values  Number of Null Values   
10                                         [1.0, 0.0]                      0  \
1                                          [1.0, 0.0]                      0   
2                                          [1.0, 0.0]                      0   
3                                          [1.0, 0.0]                      0   
18                                         [0.0, 1.0]                      0   
5                                          [1.0, 0.0]                      0   
6                                          [0.0, 1.0]                      0   
7                                          [0.0, 1.0]                      0   
8                                          [0.0, 1.0]                      0   
9                                          [0.0, 1.0]                      0   
17                                         [1.0, 0.0]                      0   
11                                         [0.0, 1.0]                      0   
12                                         [1.0, 0.0]                      0   
13                                         [0.0, 1.0]                      0   
0                                     [0.0, 2.0, 1.0]                      0   
14                          [5.0, 3.0, 2.0, 4.0, 1.0]                      0   
20                     [4.0, 6.0, 3.0, 5.0, 2.0, 1.0]                      0   
21           [3.0, 1.0, 8.0, 6.0, 4.0, 7.0, 2.0, 5.0]                      0   
19  [9.0, 7.0, 11.0, 10.0, 8.0, 13.0, 4.0, 6.0, 2....                      0   
15  [18.0, 0.0, 30.0, 3.0, 5.0, 15.0, 10.0, 6.0, 2...                      0   
16  [15.0, 0.0, 30.0, 2.0, 14.0, 28.0, 7.0, 20.0, ...                      0   
4   [40.0, 25.0, 28.0, 27.0, 24.0, 30.0, 34.0, 26....                      0   

    Proportion of Null Values    dtype  
10                        0.0  float64  
1                         0.0  float64  
2                         0.0  float64  
3                         0.0  float64  
18                        0.0  float64  
5                         0.0  float64  
6                         0.0  float64  
7                         0.0  float64  
8                         0.0  float64  
9                         0.0  float64  
17                        0.0  float64  
11                        0.0  float64  
12                        0.0  float64  
13                        0.0  float64  
0                         0.0  float64  
14                        0.0  float64  
20                        0.0  float64  
21                        0.0  float64  
19                        0.0  float64  
15                        0.0  float64  
16                        0.0  float64  
4                         0.0  float64